xapian-applications/omega/weight.cc

```diff
@@ -23,267 +24,17 @@
-		return;
-	    }
-	    if (C_isspace(*p)) {
-		double k;
-		if (!double_param(&p, &k))
-		    parameter_error("Parameter is invalid", scheme);
-		if (*p)
-		    parameter_error("Extra data after parameter", scheme);
-		enq.set_weighting_scheme(Xapian::IneB2Weight(k));
-		return;
-	    }
-	}
-
-	if (startswith(scheme, "bb2")) {
-	    const char *p = scheme.c_str() + 3;
-	    if (*p == '\0') {
-		enq.set_weighting_scheme(Xapian::BB2Weight());
-		return;
-	    }
-	    if (C_isspace(*p)) {
-		double k;
-		if (!double_param(&p, &k))
-		    parameter_error("Parameter is invalid", scheme);
-		if (*p)
-		    parameter_error("Extra data after parameter", scheme);
-		enq.set_weighting_scheme(Xapian::BB2Weight(k));
-		return;
-	    }
-	}
-
-	if (startswith(scheme, "dlh")) {
-	    const char *p = scheme.c_str() + 3;
-	    if (*p == '\0') {
-		enq.set_weighting_scheme(Xapian::DLHWeight());
-		return;
-	    }
-	    if (C_isspace(*p)) {
-		throw "No parameters are required for DLH";
-	    }
-	}
-
-	if (startswith(scheme, "pl2")) {
-	    const char *p = scheme.c_str() + 3;
-	    if (*p == '\0') {
-		enq.set_weighting_scheme(Xapian::PL2Weight());
-		return;
-	    }
-	    if (C_isspace(*p)) {
-		double k;
-		if (!double_param(&p, &k))
-		    parameter_error("Parameter is invalid", scheme);
-		if (*p)
-		    parameter_error("Extra data after parameter", scheme);
-		enq.set_weighting_scheme(Xapian::PL2Weight(k));
-		return;
-	    }
-	}
-
-	if (startswith(scheme, "dph")) {
-	    const char *p = scheme.c_str() + 3;
-	    if (*p == '\0') {
-		enq.set_weighting_scheme(Xapian::DPHWeight());
-		return;
-	    }
-	    if (C_isspace(*p)) {
-		throw "No parameters are required for DPH";
-	    }
-	}
-
-	if (startswith(scheme, "lm")) {
-	    const char *p = scheme.c_str() + 2;
-	    if (*p == '\0') {
-		enq.set_weighting_scheme(Xapian::LMWeight());
-		return;
-	    }
-	    if (C_isspace(*p)) {
-		double param_log = 0;
-		Xapian::Weight::type_smoothing type = Xapian::Weight::TWO_STAGE_SMOOTHING;
-		double smoothing1 = 0.7;
-		double smoothing2 = 2000;
-		if (!double_param(&p, &param_log))
-		    parameter_error("Parameter 1 (log) is invalid", scheme);
-		if (*p && !type_smoothing_param(&p, &type))
-		    parameter_error("Parameter 2 (smoothing_type) is invalid", scheme);
-		if (*p && !double_param(&p, &smoothing1))
-		    parameter_error("Parameter 3 (smoothing1) is invalid", scheme);
-		if (*p && !double_param(&p, &smoothing2))
-		    parameter_error("Parameter 4 (smoothing2) is invalid", scheme);
-		if (*p)
-		    parameter_error("Extra data after parameter 4", scheme);
-		Xapian::LMWeight wt(param_log, type, smoothing1, smoothing2);
-		enq.set_weighting_scheme(wt);
-		return;
-	    }
-	}
-
-	if (scheme == "coord") {
-	    enq.set_weighting_scheme(Xapian::CoordWeight());
-	    return;
-	}
-
-	if (scheme != "bool") {
-	    throw "Unknown $opt{weighting} setting: " + scheme;
-	}
+	enq.set_weighting_scheme(*(Xapian::Weight::create(scheme)));
```

Comment:
Agreed. Should've been careful here. I think after passing the object, it can simply be deleted as `
delete Xapian::Weight::create(scheme);` in the next line.